src/main/kotlin/com/openlattice/data/storage/DataDeletionService.kt

```diff
@@ -0,0 +1,434 @@
+
+        val edgeEntitySetIds = if (entityKeyIds.isPresent) {
+            dgm.getEdgeEntitySetsConnectedToEntities(entitySetId, entityKeyIds.get())
+        } else {
+            dgm.getEdgeEntitySetsConnectedToEntitySet(entitySetId)
+        }
+
+        val edgeEntitySets = edmService.getEntitySetsAsMap(edgeEntitySetIds).values
+        val nonAuditEdgeEntitySets = edgeEntitySets.filter { !it.flags.contains(EntitySetFlag.AUDIT) }
+        val entityTypesById = edmService.getEntityTypesAsMap(edgeEntitySets.map { it.entityTypeId }.toSet())
+
+        if (!skipAuthChecks) {
+            authorizeAssociations(entitySetId, nonAuditEdgeEntitySets, entityTypesById, deleteType, principals)
+        }
+
+        /* Authorizations successful. Return map from entity set ids to property types */
+
+        val allPropertyTypeIds = entityTypesById.values.flatMap { it.properties }.toSet()
+        val propertyTypesById = edmService.getPropertyTypesAsMap(allPropertyTypeIds)
+
+
+        /* Hard deletes will authorize deletion of all audit property types, while soft deletes will authorize none. */
+        val entitySetsToDeleteFrom = if (deleteType == DeleteType.Hard) edgeEntitySets else nonAuditEdgeEntitySets
+
+        return entitySetsToDeleteFrom.map { it.id }.associateWith {
+            entityTypesById.getValue(it).properties.associateWith { ptId -> propertyTypesById.getValue(ptId) }.toMap()
+                    .plus(IdConstants.ID_ID.id to PostgresMetaDataProperties.ID.propertyType)
+        }.toMap()
+    }
+
+    private fun authorizeAssociations(
+            entitySetId: UUID,
+            edgeEntitySets: List<EntitySet>,
+            entityTypesById: Map<UUID, EntityType>,
+            deleteType: DeleteType,
+            principals: Set<Principal>
+    ) {
+        val requiredPermissions = getRequiredPermissions(deleteType)
+
+        val accessChecks = mutableSetOf<AccessCheck>()
+
+        /* Ignore audit entity sets when determining delete or clear permissions */
+        edgeEntitySets.forEach {
+            accessChecks.add(AccessCheck(AclKey(it.id), requiredPermissions))
+
+            val propertyTypeIds = entityTypesById.getValue(it.entityTypeId).properties
+            propertyTypeIds.forEach { propertyTypeId ->
+                accessChecks.add(AccessCheck(AclKey(it.id, propertyTypeId), requiredPermissions))
+            }
+        }
+
+        val unauthorizedAclKeys = authorizationManager.accessChecksForPrincipals(accessChecks, principals)
+                .filter { it.permissions.values.contains(false) }
+                .map { it.aclKey }
+                .collect(Collectors.toSet())
+
+        if (unauthorizedAclKeys.isNotEmpty()) {
+            throw ForbiddenException("Unable to delete from entity set $entitySetId: missing required permissions " +
+                    "on associations for AclKeys $unauthorizedAclKeys")
+        }
+
+    }
+
+    private fun getAuthorizedPropertyTypesForDelete(
```

Comment:
Just a minor thing, but shouldn't we do the authorization checks in the controllers?

Reply:
Theoretically yes, although the auth logic is pretty deep into the delete logic in multiple spots, and moving it into the controller would involve repeating a ton of logic in multiple places (which led to copy/paste and maintenance errors in our current implementation) as well as duplicating calls unnecessarily, so it seemed safer to me to add an auth-aware delete function in DataDeletionService.

Reply:
👍